staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go

```diff
@@ -300,3 +307,61 @@ func (le *LeaderElector) maybeReportTransition() {
 		go le.config.Callbacks.OnNewLeader(le.reportedLeader)
 	}
 }
+
+type LeaderChecker interface {
+	healthz.HealthzChecker
+	SetLeaderElection(le *LeaderElector)
+}
+
+type timeNow func() time.Time
+
+type leaderChecker struct {
+	c           *apiserverconfig.LeaderElectionConfiguration
+	pointerLock sync.Mutex
+	le          *LeaderElector
+	timeNow
+}
+
+func (l *leaderChecker) Name() string {
+	return "leaderElection/" + l.c.ResourceLock
+}
+
+func (l *leaderChecker) Check(req *http.Request) error {
+	if !l.c.LeaderElect {
+		return nil
+	}
+	l.pointerLock.Lock()
+	defer l.pointerLock.Unlock()
+	if l.le == nil {
+		return nil
+	}
+	if !l.le.IsLeader() {
+		// Currently not concerned with the case that we are hot standby
+		return nil
+	}
+	now := l.timeNow()
+	// If we are more than 20 seconds after the lease durection that is past the timeout on the lease renew.
+	// Time to start reporting ourselves as unhealthy. We should have died but conditions like deadlock can
+	// prevent this. (See #70819)
+	expiry := l.le.observedTime.Add(l.le.config.LeaseDuration).Add(time.Second * 20)
```

Comment:
I'm a little conflicted about a hardcoded 20 second timeout. I think not every user of this library will exit like controller manager does, so maybe it should be configurable?

Reply:
Good point. I have pushed it as configurable up one level. However I'm not sure I want to push it all the way to the CLI just yet.